sopel/plugins/capabilities.py

```diff
@@ -0,0 +1,312 @@
+        """Tell if a capability request is requested.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._requested
+
+    def is_acknowledged(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is acknowledged.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._acknowledged
+
+    def is_denied(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is denied.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._denied
+
+    # register, request, resume, acknowledge, deny, etc.
+
+    def register(self, plugin_name: str, request: capability) -> None:
+        """Register a capability ``request`` for ``plugin_name``.
+
+        :param request: the capability request to register for later
+        """
+        self._registered.setdefault(request.cap_req, {})[plugin_name] = (
+            request, False,
+        )
+        LOGGER.debug('Capability Request registered: %s', str(request))
+
+    def request_available(
+        self,
+        bot: Sopel,
+        available_capabilities: Iterable[str],
+    ) -> None:
+        """Request available capabilities.
+
+        :param bot: the bot instance used to send capability requests
+        :param available_capabilities: available capabilities
+
+        This sends ``CAP REQ`` command for requests that can be made, i.e.
+        all the requested capabilities (with or without prefix) must be
+        available for Sopel to send the request.
+
+        Requests made are stored as requested, other are ignored.
+        """
+        # Requesting capabilities when they are available
+        capabilities = set(available_capabilities)
+        for cap_req, plugin_requests in self._registered.items():
+            # request only if all are available
+            cap_req_enables = {
+                cap.lstrip('-')
+                for cap in cap_req
+            }
+            text = ' '.join(cap_req)
+            if cap_req_enables <= capabilities:
+                self._registered[cap_req] = {
+                    plugin_name: (handler_info[0], False)
+                    for plugin_name, handler_info in plugin_requests.items()
+                }
+                self._requested.add(cap_req)
+                LOGGER.debug('Capability negotiation request: "%s".', text)
+                bot.write(('CAP', 'REQ'), text=text)
+            else:
+                LOGGER.debug(
+                    'Unable to negotiate capability request: "%s".', text,
+                )
```

Comment:
Am I missing where the callback for a request that is never sent gets notified that the capability(/ies) asked for were essentially NAK'd due to one or more not being available?

I'm not certain it's really more efficient to proactively avoid sending REQs that will obviously be NAK'd like this. Wouldn't it be easier to send the request anyway and let the server send a NAK that would propagate through the callbacks as normal?

Reply:
Well, that's the point of #972 isn't it?

Reply:
Ultimately I don't really care whether the REQ actually gets sent. I remember #972 (and that was one reason this rewrite is even happening).

I just didn't see any other code path that would give the requester feedback when a request is simply not sent due to one or more of the capabilities in it not being advertised, so suggested the easy way out (send the REQ anyway and let the NAK propagate naturally through the paths I did see).

Reply:
I understand that. In my opinion, Plugins should react to ACK and NAK only if it is relevant to the negotiation phase. If it isn't, the rules/commands/event handlers should look at the bot's state directly, and not some derived value from the ACK/NAK handler.

What I mean is that (and I may need to change the doc to explain that further/properly) is that instead of doing this:

```python
@plugin.capability('cap1')
def cap1_handler(cap_req, bot, acknowledged):
    if acknowledged:
        bot.memory['my_plugin_cap1'] = True
    else:
        bot.memory['my_plugin_cap1'] = False

@plugin.rule('something'):
def rule_something(bot, trigger):
    if bot.memory['my_plugin_cap1']:
        # do something
```

They should do this:

```python
CAP_CAP1 = plugin.capability('cap1')

@plugin.rule('something'):
def rule_something(bot, trigger):
    if bot.capabilities.is_enabled('cap1'):
        # do something
```

The difference is on how the bot's state is verified: in the first version, a proxy variable is used to know the state (the `bot.memory` flag), and a proxy needs to be synchronized. In the second version, the current bot's state is always used, there is no need for any sort of synchronization, so less risk of errors.